### ISSUE TITLE
Key sometimes not dropping from enemy

### ISSUE THREAD

**Issue by ChrisNZL** [ChrisNZL/Tallowmere2#432]
Reported in 0.0.17a.

Seems issue #168 has returned. Opening as a new issue since the previous issue was a long time ago.

Feedback ID: 0.0.17a_20191213-5VPXY, savefileID 11...
- Savegame from disk at start of room: Debug Panel says there is a key on a monster as expected.

- Mid-romp savegame from RAM after enemies have been killed: Debug Panel says there is no key anywhere, and the eastern door is still locked. No key in the home room, nor previous room.

My attempts to reproduce this have been unsuccessful. Need to investigate further.

---

Still unable to reproduce this in 0.1.0. Might end up needing to perform a polling check to spawn a key if a key is supposed to spawn but does not happen 😟

**Comment by ChrisNZL** [comment]
Seems to be an issue again, especially when playing online with 0.1.3.

**Comment by ChrisNZL** [comment]
Still happening (rarely) in 0.1.4b.

**Comment by ChrisNZL** [comment]
For v0.1.4c, I am adding two classes:

1. **`QueuedCreatureSlayer`**
   - I suspect Unity does not reliably create/instantiate things within fixed time steps 100% of the time when dealing with collisions from `Projectile.FixedUpdate`, `Projectile.OnCollisionEnter2D`, `TrapCollider.OnCollisionEnter2D`, and/or `TrapCollider.OnCollisionStay2D`.

   - As such, using Unity's `Time.inFixedTimeStep` check, the `Creature.Slay` method (which is typically called from `CreatureStruct.ReceiveDamage`) will now add data to `QueuedCreatureSlayer` to slay a creature after a moment during a regular `Update` interval – just to see if fixed timing is the issue or not.

2. **`ImportantGibPoller`**
   - Currently does nothing, but will be developed further if this issue persists. Saving this is a last resort because polling is potentially expensive.

**Comment by ChrisNZL** [comment]
Re-opening. Reported again in 0.1.6a.

Might need to flesh out **`ImportantGibPoller`**, or revamp the system to have "drop important things on death" be stored in room data, rather than on individual creatures.

Still am puzzled as to why things do not spawn 100% of the time.

**Comment by ChrisNZL** [comment]
Feedback ID: 20201004-VWMCN

v0.1.7

Missing key was reported.

Log showed:

```
5:18:55, Frame 164583, LOG »  ImportantGibPoller.OnGibSpawned: creatureNetID: Master134. gibID: Key_Yellow. gibStatus: Spawned. importantGib_instantiated: Treasure_Key_Yellow(Clone)
5:18:55, Frame 164583, LOG »  ImportantGibPoller.OnGibCollected: creatureNetID: Master134. gibID: Key_Yellow. gibStatus: Collected. importantGib_instantiated: null
```

Key was being spawned _and_ collected on the same frame.

Looking at how the key objects were put together, and looking at my code, I've identified:

- There was one positional sub-component that was tagged incorrectly on the keys – a Transform was using a Layer that is normally used for gold and hearts. So I think there was a one-frame possibility where the physics system would snag the collider on this sub-component that was tagged incorrectly, thinking it was a gold or heart -- but this would only happen if the physics tick and the game tick lined up on exactly the same frame, which was pretty rare. I've adjusted the culprit component's layer now.

- Additionally, the Key's `CanBeCollectedByCreature` method was always returning `true` – I've adjusted this method now to require checking how gold and keys can be collected – either via automatic radius area, or via pressing the Interact action, and checking against its type appropriately. (gold and keys sort of use the same underlying system)
